File: test/integration/InitializeEquivalenceDriver.py
"""The default is the dataclass; the specification's definition is deserialising nothing over it.

Usage: driver.py <generated root> <package>
"""
import importlib
import sys

sys.path.insert(0, sys.argv[1])
PACKAGE = sys.argv[2]


def load(module, name):
    return getattr(importlib.import_module(f"{PACKAGE}.{module}"), name)


def check(name, cls):
    a = cls()
    b = cls()
    consumed = b._deserialize_from(memoryview(b""))
    wa = bytearray(4096)
    wb = bytearray(4096)
    na = a._serialize_into(memoryview(wa))
    nb = b._serialize_into(memoryview(wb))
    same = consumed >= 0 and na >= 0 and na == nb and wa[:na] == wb[:nb]
    print(f"{name:<40} {'same' if same else 'DIFFER'} ({na} bytes)")
    return same


ok = True
ok = check("uavcan.node.Heartbeat", load("uavcan.node.heartbeat_1_0", "Heartbeat")) and ok
ok = check("uavcan.metatransport.can.Frame", load("uavcan.metatransport.can.frame_0_2", "Frame")) and ok
ok = check("uavcan.primitive.array.Real32", load("uavcan.primitive.array.real32_1_0", "Real32")) and ok
ok = check("uavcan.node.port.SubjectIDList", load("uavcan.node.port.subject_id_list_1_0", "SubjectIDList")) and ok
ok = check("uavcan.pnp.NodeIDAllocationData", load("uavcan.pnp.node_id_allocation_data_2_0", "NodeIDAllocationData")) and ok
ok = check("uavcan.diagnostic.Record", load("uavcan.diagnostic.record_1_1", "Record")) and ok
ok = check("uavcan.time.SynchronizedTimestamp", load("uavcan.time.synchronized_timestamp_1_0", "SynchronizedTimestamp")) and ok
sys.exit(0 if ok else 1)

File: test/integration/InitializeEquivalenceDriver.ts
// The default is the factory; the spec's definition is deserialising nothing over it.
import { makeHeartbeat, deserializeHeartbeatFrom, serializeHeartbeatInto } from "./uavcan/node/heartbeat_1_0";
import { makeFrame, deserializeFrameFrom, serializeFrameInto } from "./uavcan/metatransport/can/frame_0_2";
import { makeReal32, deserializeReal32From, serializeReal32Into } from "./uavcan/primitive/array/real32_1_0";
import { makeSubjectIDList, deserializeSubjectIDListFrom, serializeSubjectIDListInto } from "./uavcan/node/port/subject_id_list_1_0";
import { makeNodeIDAllocationData, deserializeNodeIDAllocationDataFrom, serializeNodeIDAllocationDataInto } from "./uavcan/pnp/node_id_allocation_data_2_0";
import { makeRecord, deserializeRecordFrom, serializeRecordInto } from "./uavcan/diagnostic/record_1_1";
import { makeSynchronizedTimestamp, deserializeSynchronizedTimestampFrom, serializeSynchronizedTimestampInto } from "./uavcan/time/synchronized_timestamp_1_0";

let failures = 0;
function check<T>(name: string, make: () => T, deser: (o: T, b: Uint8Array) => number, ser: (o: T, b: Uint8Array) => number): void {
  const a = make();
  const b = make();
  const db = deser(b, new Uint8Array(0));
  const wa = new Uint8Array(4096);
  const wb = new Uint8Array(4096);
  const na = ser(a, wa);
  const nb = ser(b, wb);
  const same = db >= 0 && na >= 0 && na === nb && wa.subarray(0, na).every((v, i) => v === wb[i]);
  console.log(`${name.padEnd(40)} ${same ? "same" : "DIFFER"} (${na} bytes)`);
  if (!same) { failures += 1; }
}

check("uavcan.node.Heartbeat", makeHeartbeat, deserializeHeartbeatFrom, serializeHeartbeatInto);
check("uavcan.metatransport.can.Frame", makeFrame, deserializeFrameFrom, serializeFrameInto);
check("uavcan.primitive.array.Real32", makeReal32, deserializeReal32From, serializeReal32Into);
check("uavcan.node.port.SubjectIDList", makeSubjectIDList, deserializeSubjectIDListFrom, serializeSubjectIDListInto);
check("uavcan.pnp.NodeIDAllocationData", makeNodeIDAllocationData, deserializeNodeIDAllocationDataFrom, serializeNodeIDAllocationDataInto);
check("uavcan.diagnostic.Record", makeRecord, deserializeRecordFrom, serializeRecordInto);
check("uavcan.time.SynchronizedTimestamp", makeSynchronizedTimestamp, deserializeSynchronizedTimestampFrom, serializeSynchronizedTimestampInto);
if (failures !== 0) {
  throw new Error(`${failures} default(s) differ from the specification's`);
}
